### ISSUE TITLE
Specifying that input document must be framed or compacted in algorithm?

### ISSUE THREAD

**Issue by cwebber** [w3c-dvcg/ld-signatures#12]
Best I can tell to actually implement the appropriate signature/verification algorithms you need to compact or frame the input document, but that isn't specified.  Maybe it's obvious and doesn't need to be?  Maybe it should be spelled out though, especially if framing, so the person implementing can properly find the "root" of the graph being signed.

**Comment by msporny** [comment]
> Best I can tell to actually implement the appropriate signature/verification algorithms you need to compact or frame the input document, but that isn't specified. Maybe it's obvious and doesn't need to be? 

Hmm, there seems to be some confusion here... Step w3c-ccg/ld-signatures#2 of https://w3c-dvcg.github.io/ld-signatures/#signature-algorithm states:

_Generate a canonicalized document by canonicalizing document according to a canonicalization algorithm (e.g. the GCA2015 [RDF-DATASET-NORMALIZATION] algorithm)._

So, you take your JSON-LD and go to RDF through whatever means you need. Sometimes you expand and then go straight to RDF... but I can't think of a case where you would want to compact or frame. Maybe @dlongley knows of a case where you would?

> Maybe it should be spelled out though, especially if framing, so the person implementing can properly find the "root" of the graph being signed.

Hmm... there are many "root"s, which is why you need the RDF Dataset Canonicalization Algorithm... framing doesn't help you out there. From what I remember, framing is not necessary for RDF Dataset Canonicalization, so you may be going down the wrong path here.

**Comment by cwebber** [comment]
> Hmm... there are many "root"s, which is why you need the RDF Dataset Canonicalization Algorithm... framing doesn't help you out there. From what I remember, framing is not necessary for RDF Dataset Canonicalization, so you may be going down the wrong path here.

You are right in that when the document is actually signed, it's canonicalized in an "ordered graph soup", and that doesn't have a root node.  What I was talking about was _finding_ which signature we were checking against, since an input document of n-quads may have multiple nodes with signatures, and we can't just pluck the signature off the top object like we can in json-ld, but framing would reveal what node we meant to check the signature of... is that not the case?  I was talking about framing to find that particular `signature` value to verify against, or at least compacting if already in json-ld form to make sure that `signature` is really `https://w3id.org/security#signature`.

I guess the mistake I'm making is interpreting these as step-by-step recipes that you can type in and things pretty much work (amazingly, the json-ld expander pretty much worked that way).  I suppose that an implementor might not be using json-ld at all in this algorithm, at which case they may be taking in graph soup of quads and having an argument for which id to take the signature from, or such.

**Comment by msporny** [comment]
> framing would reveal what node we meant to check the signature of... is that not the case?

Ah, now I see what you're saying. Framing is specific to JSON-LD, which is why I thought you were talking about JSON-LD. So, no, you can't use Framing in something like say... TURTLE or NQuads. Basically, however you go about it, the solution isn't Framing.

If implementers are not using JSON-LD, then they need to figure out a way for the developer to say "check the signature on node X", where X will be a URI (or canonicalized bnode name). I would imagine in those cases that they'd say "specify the URI of the subject who'se signature you'd like to check...".

This, of course, assumes that you're checking a signature that's contained IN the graph, and that's always been a hacky way to do things. We've picked this hacky way because most developers would freak if they had to understand graphs and signatures on graphs.

That said, we've hit this problem with the Verifiable Claims work and will need to extend JSON-LD so that there is named graph support in order to ensure that developers don't need to use "@graph" everywhere in Verifiable Claims. It's on our to-do list.

So, the short answer to your question is that we'll need to write some text in the spec to say that the signature to check is either assumed or directly specified when checking signature. In JSON-LD you can assume the signature object (it's at the top-level)... in NQuads, you can't, and so the developer will have to specify which signature object to check.
